cli-plugins/manager/plugin.go

```diff
@@ -0,0 +1,145 @@
+}
+
+// NewPluginError creates a new PluginError, analogous to
+// errors.Errorf.
+func NewPluginError(msg string, args ...interface{}) error {
+	return &PluginError{cause: errors.Errorf(msg, args...)}
+}
+
+// Plugin represents a potential plugin with all it's metadata.
+type Plugin struct {
+	Metadata
+
+	Name string `json:",omitempty"`
+	Path string `json:",omitempty"`
+
+	// Err is non-nil if the plugin failed one of the candidate tests.
+	Err error `json:",omitempty"`
+
+	// ShadowedPaths contains the paths of any other plugins which this plugin takes precedence over.
+	ShadowedPaths []string `json:",omitempty"`
+}
+
+// NewPlugin determines if the given candidate is valid and returns a
+// Plugin.  If the candidate fails one of the tests then `Plugin.Err`
+// is set, and is always a `PluginError`, but the `Plugin` is still
+// returned with no error. An error is only returned due to a
+// non-recoverable error.
+func NewPlugin(c Candidate, rootcmd *cobra.Command) (Plugin, error) {
+	path := c.Path()
+	if path == "" {
+		return Plugin{}, errors.New("plugin candidate path cannot be empty")
+	}
+
+	// The candidate listing process should have skipped anything
+	// which would fail here, so there are all real errors.
+	fullname := filepath.Base(path)
+	if fullname == "." {
+		return Plugin{}, errors.Errorf("unable to determine basename of plugin candidate %q", path)
+	}
+	if runtime.GOOS == "windows" {
+		exe := ".exe"
+		if !strings.HasSuffix(fullname, exe) {
+			return Plugin{}, errors.Errorf("plugin candidate %q lacks required %q suffix", path, exe)
+		}
+		fullname = strings.TrimSuffix(fullname, exe)
+	}
+	if !strings.HasPrefix(fullname, NamePrefix) {
+		return Plugin{}, errors.Errorf("plugin candidate %q does not have %q prefix", path, NamePrefix)
+	}
+
+	p := Plugin{
+		Name: strings.TrimPrefix(fullname, NamePrefix),
+		Path: path,
+	}
+
+	// Now apply the candidate tests, so these update p.Err.
+	if !pluginNameRe.MatchString(p.Name) {
+		p.Err = NewPluginError("plugin candidate %q did not match %q", p.Name, pluginNameRe.String())
+		return p, nil
+	}
+
+	if rootcmd != nil {
+		for _, cmd := range rootcmd.Commands() {
+			// Ignore conflicts with commands which are
+			// just plugin stubs (i.e. from a previous
+			// call to AddPluginCommandStubs).
+			if p := cmd.Annotations[CommandAnnotPlugin]; p == "true" {
+				continue
+			}
+			if cmd.Name() == p.Name || cmd.HasAlias(p.Name) {
+				p.Err = NewPluginError("plugin duplicates builtin command")
+				return p, nil
+			}
+		}
+	}
+
+	// We are supposed to check for relevant execute permissions here. Instead we rely on an attempt to execute.
+	meta, err := c.Metadata()
+	if err != nil {
+		p.Err = WrapAsPluginError(err, "failed to fetch metadata")
+		return p, nil
+	}
+
+	if err := json.Unmarshal(meta, &p.Metadata); err != nil {
+		p.Err = WrapAsPluginError(err, "invalid metadata")
+		return p, nil
+	}
+
+	if p.Metadata.SchemaVersion != "0.1.0" {
```

Comment:
You can use a library like `masterminds/semver`, check this to get inspiration https://github.com/docker/kamoulox-compose/blob/745afa4cf7ec811756c116c57a22470c561e6f8a/install/requirements.go#L33 😄

Reply:
As soon as there is a 0.2.0 etc that'll be useful, right now there is exactly one valid version so a string compare suffices. I've used github.com/blang/semver elsewhere too.

Reply:
Yep `github.com/blang/semver` or `github.com/masterminds/semver` whatever, we just need to check if there isn't already one used somewhere in the CLI.
👍 to do that when a `0.2.0` will be there. But then maybe change the error message to something more clearer, like ` NewPluginError("plugin SchemaVersion %q is not valid, it must be 0.1.0", p.Metadata.SchemaVersion)`, it will be easier for plugin writers to catch that ?